### PR TITLE
[#61] As a user, I can update my own profile details

### LINKED ISSUE #61: As a user, I can update my own profile details
## Why
When the users logged in the application successfully, they will be able to update their profile details with the `Edit Profile` button in the left menu header from `Home` Screen.

## Acceptance Criteria
- [ ] Get the current user profile and populate data on the text fields for the current user in the `Update Profile` screen, including the user avatar url, the username, the user bio, the user email, and the user password.
- [ ] If there is no values on the text fields, disable the `Update` button by default.
- [ ] When the users update with valid values on the text fields, enable the `Update` button and need to check to disable the button again if the users modifications are clearing all contents in the textfields.
- [ ] When the users press the `Update` button, call the API to update user profile and show a native loading indicator in the middle of the screen while doing so.
- [ ] When there is an error while updating the current user, show a temporary toast message with text: `Something went wrong. Please try again later.` and just dismiss the loading indicator.
- [ ] If the update is successful, store the new user data to the session cache, dismiss the loading indicator and go back to the previous screen.

## Resource
- Sample UX Flow:

https://user-images.githubusercontent.com/70877098/130315909-e554bf1b-c31d-4dc4-9091-ab6a2e763580.mov

### PR DESCRIPTION
**Pull request by minhnimble**
Resolved #61

## What happened

When the users logged in the application successfully, they will be able to update their profile details with the `Edit Profile` button in the left menu header from `Home` Screen.

## Insight

- [x] Get the current user profile and populate data on the text fields for the current user in the `Update Profile` screen, including the user avatar url, the username, the user bio, the user email, and the user password.
- [x] If there is no values on the text fields, disable the `Update` button by default.
- [x] When the users update with valid values on the text fields, enable the `Update` button and need to check to disable the button again if the users modifications are clearing all contents in the textfields.
- [x] When the users press the `Update` button, call the API to update user profile and show a native loading indicator in the middle of the screen while doing so.
- [x] When there is an error while updating the current user, show a temporary toast message with text: `Something went wrong. Please try again later.` and just dismiss the loading indicator.
- [x] If the update is successful, store the new user data to the session cache, dismiss the loading indicator and go back to the previous screen.
- [x] Add unit tests for multiple files related to this feature: `EditProfileViewModel` and `SideMenuViewModel `.

## Proof Of Work

https://user-images.githubusercontent.com/70877098/137072395-b593bfff-b39e-4341-9c8f-14366a94ee10.mov

Tests passed for `EditProfileViewModel`:

Screen Shot 2021-10-13 at 12 23 21

Tests passed for `SideMenuViewModel `:

Screen Shot 2021-10-13 at 12 23 07

### CODE REVIEW

File: NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileView+UIModel.swift [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+extension EditProfileView {
+
+     struct UIModel: Equatable {
+
+         let username: String
+         let email: String
+         let avatarURL: String
+         let bio: String
+
+         init(username: String = "", email: String = "", avatarURL: String = "", bio: String = "") {
```
Comment (markgravity) [inline NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileView+UIModel.swift:19]: Please chop down the parameters when it's too many 🙏

File: NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import RxSwift
+import RxCocoa
+import Combine
+import Resolver
+
+// sourcery: AutoMockable
+protocol EditProfileViewModelInput {
+
+    func didTapUpdateButton(
+        username: String, email: String, password: String, avatarURL: String, bio: String
```
Comment (markgravity) [inline NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileViewModel.swift:17]: Please chop down when the parameters are too many 🙏

Comment (minhnimble) [reply NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileViewModel.swift:17]: same here

File: NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileView+UIModel.swift (hunk 1 above)
Comment (minhnimble) [reply NimbleMedium/Sources/Presentation/Modules/EditProfile/EditProfileView+UIModel.swift:19]: According to our team's convention, we don't have to add line break for each param for the init function if line length constraint is not exceeded 🙏

File: NimbleMedium/Sources/Domain/Repositories/Auth/AuthRepositoryProtocol.swift [hunk 3]
```diff
@@ -20,7 +20,7 @@ protocol AuthRepositoryProtocol: AnyObject {
         username: String,
         email: String,
         password: String?,
-        image: String?,
-        bio: String?
+        image: String,
+        bio: String
```
Comment (suho) [inline NimbleMedium/Sources/Domain/Repositories/Auth/AuthRepositoryProtocol.swift:24]: @minhnimble I prefer update these code in backend ticket and also use a struct to reduct changes on different places
Screen Shot 2021-10-14 at 10 02 30

Comment (minhnimble) [reply NimbleMedium/Sources/Domain/Repositories/Auth/AuthRepositoryProtocol.swift:24]: @suho yeah agree, previously when implementing the backend ticket I am not aware that we need to this until playing around with the endpoint using various different type of inputs 🙏

**Review by suho — APPROVED** [review]

**Pull request merged**